lib/em-http/http_encoding.rb

```diff
@@ -51,7 +51,10 @@ def encode_request(method, uri, query, proxy)
 
       # Non CONNECT proxies require that you provide the full request
       # uri in request header, as opposed to a relative path.
-      query = uri.join(query) if proxy
+      # Don't modify the header with CONNECT proxies. It's unneeded and will
+      # cause 400 Bad Request errors with many standard setups.
+      is_connect = proxy && (proxy[:type] == :http || proxy[:type].nil?) && uri.scheme == "https"
+      query = uri.join(query) if proxy && !is_connect
```

Comment:
Sure, I like passing in connopts and will do that. I guess I wasn't sure if I permission to change up any internal interfaces. 

Not sure if your memory is correct on the apache/squid thing. For Non-connect HTTP, the full url needs to be in the request for sure since it needs to know where to setup the connection to and has no where else to get it other than parsing the http message headers. 

For connect, two requests are sent, the first "CONNECT host:port HTTP/version" message to set up the connection then the actual http request.  The header for the first message isn't  affected by this code other than fixing the nil port thing. So I don't see it being possible for this code to break apache/squid based proxies. The code change affects the header of the second message, which I don't think it should. In the connect scenario the proxy is acting as a dumb tunnel for ssl data so there's no need to rewrite headers of requests passing though.